Optimize short-circuitable folds in `MCP`

sim642 opened the pull request:
This PR is an attempt to optimize the following `MCP` domain/context/access functions:
* `equal`
* `compare`
* `leq`
* `is_top`
* `is_bot`
* `may_race`

The previous implementations used folds, which inevitably had to iterate over the entire list each time. Even though the folding functions themselves used the short-circuiting binary operators, this didn't short-circuit much. Moreover, all the inefficient `List.assoc` calls were still happening just to pass the corresponding arguments to the folding function (which in the short-circuiting case didn't even use them).
These `assoc`s are a pain because as the fold goes down the list, each corresponding `assoc` needs to go through more of the other list. Overall, this means O(n²) complexity.

This PR implements alternative short-circuitable folds for `MCP`, which abort the list iteration via local exception to short-circuit. It not only avoids the unnecessary list tail iteration, but also all the `assoc` calls that would unnecessarily happen there.

This is quite relevant because while profiling slow race warning performance that @vesalvojdani noticed, it turns out that 30% of race warning time is just spent on `MCPAccess.A.compare` while manipulating sets of accesses.
I'm speculating that this might even have general performance improvement since path-sensitivity also uses sets of `MCP.D` values.


### TODO
- [x] Benchmark.

michael-schwarz commented on src/analyses/mCPRegistry.ml:
```diff
@@ -251,24 +273,41 @@ struct
       else fold_left (fun a (n,d) -> f a n d @@ assoc n y) a x
     with Not_found -> raise (DomListBroken "binop_fold : assoc failure")
 
+  let binop_short_fold_bool f a (x:t) (y:t) =
+    let f n d1 d2 =
+      f n (assoc_dom n) d1 d2
+    in
+    try if length x <> length y
+      then raise (DomListBroken "binop_short_fold_bool : differing lengths")
+      else (
+        iter (fun (n,d) -> f n d @@ assoc n y) x;
+        a
+      )
+    with
+      | Not_found -> raise (DomListBroken "binop_short_fold_bool : assoc failure")
+      | ShortCircuitBool b -> b
+
```
Do we really need to give `binop_short_fold_bool` both in `DomListPrintable` and in `DomListLattice`?

sim642 replied:
It's for the same reason there were `binop_fold` in both before: one uses first class modules of type `(module Printable.S)` and the other of `(module Lattice.S)`.

michael-schwarz commented:
If it is really the calls to `assoc` that are biting us here, one could also think about at least reducing the complexity of `assoc_dom` calls by introducing some dedicated data structure for it that makes accesses `O(1)`.

As the list of all available analyses is typically small, does not change during analysis, and numbered starting at zero, one could think about generating a `Module option array` where the index corresponds to the number of the analysis?

sim642 commented:
> If it is really the calls to `assoc` that are biting us here, one could also think about at least reducing the complexity of `assoc_dom` calls by introducing some dedicated data structure for it that makes accesses `O(1)`.
> 
> As the list of all available analyses is typically small, does not change during analysis, and numbered starting at zero, one could think about generating a `(module Bla.S) option array` where the index corresponds to the number of the analysis?

I've been thinking about this as well, but also the fact that the whole assoc list representation in MCP might need overhauling, for example to use `Map` instead. Another possibility would be to ensure that analyses in those lists have well-defined and consistent order (e.g. sorted), so they could simply be zipped. I'll open a new issue about this.

sim642 commented:
On sv-benchmarks there's not much of a difference, maybe ~2% improvement:
![image](https://user-images.githubusercontent.com/378740/151373192-88759f56-8a7a-4e56-8e03-6d815147b943.png)
[Full results table](https://goblint.cs.ut.ee/results/59-all-fast-mcp-short-fold-after/results/results.2022-01-27_13-51-13.table.html#/scatter?toolY=1&columnY=1&scaling=Linear&results=All&regression=Linear&filter=0(0*status*(status(in(unknown,true,EXCEPTION%20(Failure),EXCEPTION,EXCEPTION%20(IntDomain.IncompatibleIKinds),EXCEPTION%20(Cilfacade.TypeOfError))))),1(0*status*(status(notIn(ERROR%20(verify),EXCEPTION%20(Stack%20overflow),OUT%20OF%20MEMORY,TIMEOUT,TIMEOUT%20(true)))))&line=1.1)

But there also isn't big multithreaded benchmarks there, so the differences aren't as pronounced as in #571.

sim642 commented:
I would merge this after #578 because there are major conflicts between the two.

sim642 commented:
I now rebased this on top of #578 and also simplified the implementations: instead of using exceptions to escape the fold, I'm now just using `for_all*` functions on lists, which also stop iterating on first `false`.

I also benchmarked this new version and it's much more impressive than the previous version.

### This compared to #578
There's an **additional ~10% speedup:**
![image](https://user-images.githubusercontent.com/378740/151809934-031c815f-9ef9-4cf3-bc6d-39e655552f48.png)

### #578 and this compared to master
Together the two give **~27% speedup:**
![image](https://user-images.githubusercontent.com/378740/151809998-e31ecc72-5017-41ac-b49a-c9b538f67386.png)

[Full results table](https://goblint.cs.ut.ee/results/62-all-fast-mcp-no-assoc-after-short-circuit/results/results.2022-01-31_14-04-22.table.html#/scatter?toolY=2&columnY=1&toolX=0&columnX=1&results=All&filter=0(0*status*(status(in(unknown,true,EXCEPTION,EXCEPTION%20(Failure),EXCEPTION%20(IntDomain.IncompatibleIKinds),EXCEPTION%20(Cilfacade.TypeOfError))))),1(0*status*(status(in(unknown,true,EXCEPTION,EXCEPTION%20(Failure),EXCEPTION%20(IntDomain.IncompatibleIKinds),EXCEPTION%20(Cilfacade.TypeOfError))))),2(0*status*(status(in(unknown,true,EXCEPTION%20(Failure),EXCEPTION%20(IntDomain.IncompatibleIKinds),EXCEPTION%20(Cilfacade.TypeOfError)))))&regression=Linear&scaling=Linear&line=2)

michael-schwarz commented:
Nice! I'm always amazed at how much time we seem to waste in non-obvious places.
